zookeeper-server/src/main/java/org/apache/zookeeper/server/Request.java

```diff
@@ -305,4 +305,25 @@ public void setException(KeeperException e) {
     public KeeperException getException() {
         return e;
     }
+
+    public void logLatency(ServerMetrics metric, String key, long currentTime) {
+        if (hdr != null) {
+            /* Request header is created by leader. If there is clock drift
+             * latency might be negative. Headers use wall time, not
+             * CLOCK_MONOTONIC.
+             */
+            long latency = currentTime - hdr.getTime();
+            if (latency > 0) {
+                if (key != null) {
+                    metric.add(key, latency);
```

Comment:
I can't understand this point.

Is key optional?
I am thinking about the new API, I would like to make it explicit in the new API what is the behavior of not using the key.

Reply:
If the metric is of type Set, then we need the key. If not, then key should be null.  For example, ACK_LATENCY is of type AvgMinMaxPercentileCounterSet, meaning we record the latency for each server and the server_id is the key. 

Maybe it's better to overload logLatency(), one with key parameter, one w/o parameter, instead of checking null. What do you think?

Also I have been thinking how to represent Set metrics using the new APIs. My intuition is that context should be the answer but we probably need a couple more APIs.

Reply:
For the new API I have a new proposal which fits the Counter Set summaries.
Please take a look here

https://github.com/apache/zookeeper/pull/854/files

This new patch covers perfectly existing model of your set of patches, it is 100% compatible

Reply:
......Maybe it's better to overload logLatency(), one with key parameter, one w/o parameter, instead of checking null. What do you think?


Yes it may be a good idea, can you try?

Reply:
> ......Maybe it's better to overload logLatency(), one with key parameter, one w/o parameter, instead of checking null. What do you think?
> 
> Yes it may be a good idea, can you try?

Ok, I will refactor the code.